### PR TITLE
Car class is made immutable

### PR DESCRIPTION
**Pull request by DmGrt**

### CODE REVIEW

File: src/main/java/core/basesyntax/Car.java [hunk 1]
```diff
@@ -1,13 +1,85 @@
...
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        List<Wheel> newWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            newWheels.add(wheel.clone());
+        }
+        this.wheels = newWheels(wheels);
+        this.engine = engine == null ? null : engine.clone();
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:21]: the same about separate private method

File: src/main/java/core/basesyntax/Car.java [hunk 2]
```diff
@@ -1,13 +1,85 @@
...
+        }
+        return newWheels;
+    }
+
+    public Engine getEngine() {
+        return engine == null ? null : engine.clone();
+    }
+
+    public Car changeColor(String color) {
+        return new Car(getYear(), color, getWheels(), getEngine());
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:45]:
```suggestion
        return new Car(year, color, wheels, engine);
```

File: src/main/java/core/basesyntax/Car.java [hunk 3]
```diff
@@ -1,13 +1,85 @@
...
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> newWheels = new ArrayList<>(wheels);
+        newWheels.add(wheel.clone());
+        return new Car(getYear(), getColor(), newWheels, getEngine());
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:51]:
```suggestion
        return new Car(year, color, newWheels, engine);
```

File: src/main/java/core/basesyntax/Car.java [hunk 4]
```diff
@@ -1,13 +1,85 @@
...
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(getYear(), getColor(), getWheels(), engine);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:55]: the same

File: src/main/java/core/basesyntax/Car.java [hunk 5]
```diff
@@ -1,13 +1,85 @@
...
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        List<Wheel> newWheels = new ArrayList<>();
+        for (Wheel wheel : wheels) {
+            newWheels.add(wheel.clone());
+        }
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:19]: do we need this for loop here?

File: src/main/java/core/basesyntax/Car.java [hunk 6]
```diff
@@ -1,13 +1,85 @@
...
+                && Objects.equals(getWheels(), car.getWheels())
+                && Objects.equals(getEngine(), car.getEngine());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getYear(), getColor(), getWheels(), getEngine());
+    }
+
+    private List<Wheel> newWheels(List<Wheel> wheels) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:78]:
```suggestion
    private List<Wheel> getCopyOfWheels(List<Wheel> wheels) {
```

File: src/main/java/core/basesyntax/Car.java [hunk 7]
```diff
@@ -1,13 +1,85 @@
...
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/Car.java:32]: let's use here method - getCopyOfWheels()

File: src/main/java/core/basesyntax/Car.java [hunk 8]
```diff
@@ -1,13 +1,81 @@
...
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof Car)) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:55]: please refactor this check in all three classes

File: src/main/java/core/basesyntax/Car.java [hunk 9]
```diff
@@ -1,13 +1,81 @@
...
+            return false;
+        }
+        Car car = (Car) o;
+        return getYear() == car.getYear()
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Car.java:59]: and don't use getters in `equals()` and `hashCode()` in all three classes

**Pull request closed without merging**